### ISSUE TITLE
Throw an error when attempting to redefine control structure keywords

### ISSUE THREAD

**Issue by brianlheim** [supercollider/supercollider#2651]
See discussion at #754 for background.

Control-flow keywords that are interpreted as bytcodes by the compiler are already impossible to override — switch is one, for instance. I propose adding a compiler error message when trying to redefine this in a class or variable, and adding syntax highlighting in the same color as `thisProcess` et al. to control structure keywords `if`, `while`, `do`, `for`, `case`, `switch`, and others as appropriate.

**Comment by jamshark70** [comment]
I think it's appropriate to do this only for keywords which the compiler inlines. I believe that's only `if`, `while`, `case` and `switch`. (But the class library depends on a non-inlined `if` method call for a linear crossfade between UGens, so don't be overzealous.)

`for` and `do` *look* like control structures, but in SC, technically they aren't. They are normal method calls that happen to perform loops.

**Comment by telephon** [comment]
Just to be sure I understand: you propose that the above methods can only be implemented by classes that currently implement them? Certainly `if` is already an exception, it is implemented for `UGen`. Of course `do` is, too. I see no special reason why you shouldn't be able to override the others. If this goes wrong, is there perhaps something wrong with inlining?

**Comment by jamshark70** [comment]
This is a tricky one -- be very *very* careful about the functional spec. I would say this is tricky enough that... please **don't** write any code until we can agree on the spec.

@telephon - point of clarification, `do` should not be part of this discussion because the compiler never inlines it. That underscores my point about being precise -- in fact, `do` is always just a normal method call. We *think* of it as a control structure, but in terms of the compiler and bytecodes, it isn't one.

```
{ 10.do { |x| x.postln } }.def.dumpByteCodes;

BYTECODES: (8)
  0   2C 0A    PushInt 10
  2   04 00    PushLiteralX instance of FunctionDef - closed
  4   B0       TailCallReturnFromFunction
  5   C2 1D    SendSpecialMsg 'do'
  7   F2       BlockReturn
```

Where a control structure is inlined, you'll see `ControlOpcode` or some form of `Jump...` in the bytecodes.

```
{ if(0.5.coin) { "yep" } { "nope" } }.def.dumpByteCodes;

BYTECODES: (12)
  0   66       PushSpecialValue 0.5
  1   0D 2C    SendSpecialUnaryArithMsgX 'coin'
  3   F8 00 04 JumpIfFalse 4  (10)
  6   41       PushLiteral "yep"
  7   FC 00 01 JumpFwd 1  (11)
 10   40       PushLiteral "nope"
 11   F2       BlockReturn
```

`UGen:if` works because a/ in the valid usage, the true and false branches are UGens -- hence the compiler can't inline, and the method call will be used, and b/ if the user supplies inlinable functions, the must-be-Boolean error is the correct outcome. So, here's a case where we can distinguish between normal `if` and the override based on arguments that are impossible to inline.

I'm not sure the same is true for `switch` -- but I also can't say it's impossible to imagine a `switch` override for some class where proper arguments for it would rule out inlining.

**Comment by jamshark70** [comment]
While I'm procrastinating on making music (ha!), I thought I'd take the control-structure cases one by one.

TL;DR -- it may very well be enough to implement a warning for `switch` only, and leave the others alone. Don't overdo it.

**Easy cases**

`while`, `loop` and `case` can inline only if the receiver is a literal function. So I think, for these, there is no need to impose any restriction. The only way to invoke a class-specific override for these is to write `(something that is not a literal function).while` and the compiler will always produce a normal method call in this case.

```
{ loop { 1.wait } }.def.dumpByteCodes;  // inlined
{ loop(x) }.def.dumpByteCodes;  // not inlined

{ while { x.isNil } { x = 1 } }.def.dumpByteCodes;  // inlined
{ var f = { x.isNil }; while(f, { x = 1 }) }.def.dumpByteCodes;  // not inlined

{ case { x.isNil } { x = 1 } }.def.dumpByteCodes;  // inlined
{ var f = { x.isNil }; case(f) { x = 1 } }.def.dumpByteCodes;  // NOT inlined
```

**Medium-tricky case: if**

`if` can't distinguish between the inlined or method-call cases using the receiver. It depends entirely on the arguments: inline if there are one or two arguments (excluding receiver) AND both are inlineable blocks.

It's potentially dangerous to implement `if` on something where it would be normal for user code to supply only one or two function arguments (though three would be OK). *But* `UGen:if` isn't a problem because the two arguments are expected not to be functions.

Also, some risk is alleviated here because, when it's inlined, the receiver must be Boolean or you get a runtime error -- no silent failures.

```
{ if(x.isNil) { "y" } { "n" } }.def.dumpByteCodes;  // inlined
{ var c = x.isNil; if(c) { "y" } { "n" } }.def.dumpByteCodes;  // inlined
{ if(x.isNil) { "y" } { "n" } { "invalid" } }.def.dumpByteCodes;  // 3 funcs, not inlined
```

**Switch**

Inlining is more permissive than `if`: like `if`, the compile-time receiver can be anything, but you can have any number of arguments, and every other argument could be a literal (not only functions).

This is probably why the issue was logged for `switch`, but nobody thought of it as a problem for the other inlining control structures: it's easier to think of an override which, in user code, would accidentally trigger inlining. And, the runtime class of the receiver would not throw an error, so you can get silent failures with `switch` (and that's unique to this method selector).

```
{ switch(x) { 1 } { x = 2 } }.def.dumpByteCodes;  // inlined
{ switch(x, 1, { x = 2 }) }.def.dumpByteCodes;  // **inlined**
{ switch(x, 1, 2) }.def.dumpByteCodes;  // block arg not a func: not inlined
```

**Comment by brianlheim** [comment]
After some thought about this, I've realized that what I originally meant to say was "turn control structure methods into reserved keywords." That is to say they would not only be highlighted like `thisProcess` etc., but they would also not be implemented on _any_ class. Obviously this is not something we can do in 3.9, or probably want to do in SC 3 at all, because in order to do it right it would have to be accompanied by some pretty major rewrites of older language code.

To be honest I'm not very familiar with the sclang compiler, so this was a rather naïve way to start the discussion, but it does seem to have gotten some interesting facts out in the open!

**Comment by telephon** [comment]
Also conceptually, control structure methods are just normal messages sent to objects. If the compiler finds a way to optimize it, good. But the basic logic underlying the whole OOP system is that *only the receiver of a message matters for its meaning*. 

In some cases, like binary math ops, you perform a dispatch on the first argument to decide the correct call.

We have some cases where optimization causes inconsistencies, because the primitives don't know enough about the class library. We can think about how to improve that (I'm not sure though how).

**Comment by jamshark70** [comment]
> After some thought about this, I've realized that what I originally meant to say was "turn control structure methods into reserved keywords."

Elaborating a bit on Julian's comment -- this would move the SuperCollider language a bit further away from its SmallTalk roots. Something would be gained -- hard-and-fast meanings for certain keywords -- but something would be lost too.

> ... they would also not be implemented on any class...

Perhaps underappreciated here is part of SmallTalk's conceptual elegance: **you don't need control structures at all.** This is such a radical notion that it's easy to misunderstand, and I think Brian's last comment comes from a viewpoint where it's impossible to imagine a programming language that does away with hardwired conventional control structures -- that is, it assumes that control structures *should or must* be handled as special mechanisms.

Control structures are branching and looping.

**Branching** can be handled entirely by polymorphism: https://github.com/supercollider/supercollider/blob/master/SCClassLibrary/Common/Core/Boolean.sc#L37 ff. If you don't get this, read it again. It's mind blowing. *There is nothing special about `if`.* SC inlines certain `if` forms as a performance optimization, but that's not strictly necessary. The point being, coming from Pascal / C / Java, you're conditioned to believe that `if` has a special meaning and must be handled specially. Nope.

**Looping** could be handled with recursion. SC inserts special bytecodes for `Object:while`, but in theory, if tail call optimization could prevent the stack from growing out of control, there is no reason why `while` could not look like this:

```
	whileRecurs { |body|
		if(this.value) {
			body.value;
			this.whileRecurs(body);
		}
	}
```

It happens, in SC, that you don't get the benefit of tail call optimization on `this.whileRecurs(body)` so long loops would eventually blow out the stack -- hence the "special byte codes." The point is, considering only algorithmic requirements, polymorphism plus first-class functions completely eliminates the need for C-style control structures. **Completely.**

One of the things that I *like* about the SuperCollider language is that it *isn't* Java. I can understand Brian's argument in the terms "having introduced *some* optimizations for control structures, SC could be more consistent and enforce them in every case." But this is at the cost of diversity in the computer science ecosystem.

Sorry, I know I've made a few long posts on this, but I feel strongly about it.

**Comment by jamshark70** [comment]
Oh... you *can* get tail call optimization for `while` by using ^ for early exit. Ha!

```
	whileRecurs { |body|
		if(this.value.not) { ^this };
		body.value;
		^this.whileRecurs(body);
	}
```

**Comment by brianlheim** [comment]
I want to continue discussion on this, and I will post more later, but I'm closing because even thinking about the big picture now, even I don't want this change.

**Comment by jamshark70** [comment]
Thanks. I have one more comment, didn't realize it while I was writing the other stuff. I had a sense from the beginning that SC is more "fun" than more conventional languages, and this discussion finally clarified why.

In C/Java, the language design dictates control structures -- in a sense, branching and looping "come from God." There's an authoritarian element to it: it's not even that some keywords are more equal than others, but rather that some keywords are inviolable and strictly enforced.

In the kind of "pure" SmallTalk I was mentioning above, the language design *creates the potential* for looping and branching but enforces nothing. It's all up to the class library: literally, all keywords *are* created equal. Freedom is messy, but it's also more joyful, and even though I've complained a lot over the last year or two, I think this is one reason why I stick with it -- because it's one small corner where egalitarianism works.

That's probably as philosophical as I will ever get about computer languages, but it explains why I'm quite serious when I say: if we lost this sort of freedom in SC, I don't know if I could stick around and I would probably switch to Haskell (another language that replaces traditional branching with a sophisticated dispatch mechanism).

**Comment by telephon** [comment]
*But* especially in the case of *if*, the error mechanism gets in the way of good design. If you try to implement a Liar:

```

Liar {
	var <>value;

	*new { |bool| ^this.newCopyArgs(bool) }

	if { arg trueFunc, falseFunc; ^value.not.if(trueFunc, falseFunc) }
	not { ^this.class.new(value) }
	&& { arg that; ^this.class.new(value.not && that) }
	|| { arg that; ^this.class.new(value.not || that) }
	and { arg that; ^this.class.new(value.not and: that) }
	or { arg that;  ^this.class.new(value.not || that) }
	nand { arg that; ^this.class.new(value.not.nand(that)) }
}

```

You can't use it in an `if` statement (`ERROR: Non Boolean in test.`). To make the backend more compatible with sclang's Smalltalk structure, it would be good if the test would follow some more flexible prinicple (e.g. a message implementation `isBoolean`).

**Comment by jamshark70** [comment]
Yes, inlining compromises the freedom. I'm not especially bothered by a little impurity, though.

As you note, it's fairly easy to address for methods that expect Boolean. `switch` remains a stubborn case, and uniquely so.
